File: chatly-backend/src/main/java/com/chatly/model/mongo/Message.java
package com.chatly.model.mongo;

import com.chatly.model.enums.MessageStatus;
import com.chatly.model.enums.MessageType;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@Document(collection = "messages")
@CompoundIndex(def = "{'conversationId': 1, 'createdAt': -1}")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Message {

    @Id
    private String id;

    private String conversationId;

    private String senderId;

    private String content;

    @Builder.Default
    private MessageType type = MessageType.TEXT;

    @Builder.Default
    private MessageStatus status = MessageStatus.SENT;

    private String replyToId;

    private String forwardedFromId;

    private String forwardedFromConversationId;

    @Builder.Default
    private List<Attachment> attachments = new ArrayList<>();

    @Builder.Default
    private List<ReadReceipt> readBy = new ArrayList<>();

    // Recall fields
    @Builder.Default
    private boolean recalled = false;
    private Instant recalledAt;
    private String recalledBy;

    // Edit fields
    @Builder.Default
    private boolean edited = false;
    private Instant editedAt;

    @Builder.Default
    private List<EditHistory> editHistory = new ArrayList<>();

    @Builder.Default
    private List<Reaction> reactions = new ArrayList<>();

    // Poll (for POLL type messages)
    private Poll poll;

    // Pin
    @Builder.Default
    private boolean pinned = false;
    private Instant pinnedAt;
    private String pinnedBy;

    // Priority tag (NORMAL, IMPORTANT, URGENT)
    private String priority;

    // Mentions – list of user IDs mentioned in this message ("all" for @all)
    @Builder.Default
    private List<String> mentions = new ArrayList<>();

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}

File: chatly-frontend/src/types/message.ts
import type { MessageType, ConversationResponse } from "@/types/conversation";

export type MessageStatus = "SENT" | "DELIVERED" | "READ";

export interface ReadReceipt {
    userId: string;
    readAt: string;
}

export interface Attachment {
    fileId?: string;
    kind?: "POST_PREVIEW" | "REEL_PREVIEW" | "STORY_REPLY";
    url: string;
    name?: string;
    type?: string;
    size?: number;
    durationSeconds?: number;
    // POST_PREVIEW fields
    postId?: string;
    postTitle?: string;
    postExcerpt?: string;
    postImageUrl?: string;
    postAuthorName?: string;
    postAuthorAvatarUrl?: string;
    targetUrl?: string;
    // REEL_PREVIEW fields
    reelId?: string;
    reelCaption?: string;
    reelVideoUrl?: string;
    reelAuthorName?: string;
    reelAuthorAvatarUrl?: string;
    // STORY_REPLY fields
    storyId?: string;
    storyType?: string;
    storyMediaUrl?: string;
    storyContent?: string;
    storyOwnerName?: string;
    storyOwnerAvatarUrl?: string;
}

export interface EditHistoryEntry {
    content: string;
    editedAt: string;
}

export interface Reaction {
    userId: string;
    emoji: string;
    createdAt: string;
}

export interface Poll {
    question: string;
    options: string[];
    multipleChoice: boolean;
    votes: Record<string, string[]>; // optionIndex -> userIds
    closed?: boolean;
    deadline?: string;
    anonymous?: boolean;
}

export interface LocationPayload {
    latitude: number;
    longitude: number;
    address?: string;
    mapSnapshotUrl?: string;
}

/**
 * Message returned from API – matches 100% with MessageResponse.java.
 * API order: latest first (descending), needs reverse before rendering.
 */
export interface Message {
    id: string;
    conversationId: string;
    senderId: string;
    content: string;
    type: MessageType;
    status: MessageStatus;
    replyToId: string | null;
    forwardedFromId: string | null;
    forwardedFromConversationId: string | null;
    attachments: Attachment[];
    readBy: ReadReceipt[];

    // Recall fields
    recalled: boolean;
    recalledAt: string | null;
    recalledBy: string | null;

    // Edit fields
    edited: boolean;
    editedAt: string | null;
    editHistory: EditHistoryEntry[];

    // Reactions
    reactions: Reaction[];

    // Poll
    poll?: Poll | null;

    // Location
    location?: LocationPayload | null;

    // Pin
    pinned: boolean;
    pinnedAt: string | null;
    pinnedBy: string | null;

    // Priority tag
    priority?: string | null; // "IMPORTANT" | "URGENT" | null

    // Mentions – user IDs mentioned in this message ("all" for @all)
    mentions?: string[];

    createdAt: string;
    updatedAt: string;
}

/**
 * ChatEvent – wrapper for all realtime message and group update events from WebSocket.
 */
export type ChatAction = "SEND" | "EDIT" | "RECALL" | "DELETE" | "GROUP_UPDATE" | "REACT" | "ROLE_UPDATED";
export interface ChatEvent {
    action: ChatAction;
    message?: Message;
    conversationData?: ConversationResponse;
}

/**
 * ChatUser – display information of a participant in the conversation.
 */
export interface ChatUser {
    id: string;
    displayName: string;
    username: string;
    avatarUrl?: string;
    suspended?: boolean;
    phone?: string;
    dob?: string;
    role?: "USER" | "ADMIN";
    privacy?: {
        showPhone?: boolean;
        showDob?: boolean;
    };
}
